### PR TITLE
Adds Inner http status into gql error extensions

### PR DESCRIPTION
**Pull request by scottdouglas1989**
Previously https://github.com/apollographql/router/pull/2902 added in the status code to the Error enum

This was not properly making it out into the graphql extension, because it was stripped to None in the `map_err`

This fix passes that status code forward so it gets serialized into the gql error extensions, and is available for plugins to inspect.

### CODE REVIEW

**Comment by scottdouglas1989** [comment]
Without this patch, Subgraph fetch errors end up looking like:

```
"errors": [
		{
			"message": "HTTP fetch failed from 'players': HTTP fetch failed from 'players': 429: Too Many Requests",
			"path": [],
			"extensions": {
				"status_code": Null,
				"service": "players",
				"reason": "HTTP fetch failed from 'players': 429: Too Many Requests",
				"code": "SUBREQUEST_HTTP_ERROR"
			}
		}
	]
```

With this change:

```
"errors": [
		{
			"message": "HTTP fetch failed from 'players': HTTP fetch failed from 'players': 429: Too Many Requests",
			"path": [],
			"extensions": {
				"status_code": 429,
				"service": "players",
				"reason": "HTTP fetch failed from 'players': 429: Too Many Requests",
				"code": "SUBREQUEST_HTTP_ERROR"
			}
		}
	]
```

File: apollo-router/src/query_planner/fetch.rs [hunk 1]
```diff
@@ -266,10 +266,30 @@ impl FetchNode {
...
-                reason: e.to_string(),
+            .map_err(|e| {
+                if !e.is::<FetchError>() {
+                    FetchError::SubrequestHttpError {
+                        status_code: None,
+                        service: service_name.to_string(),
+                        reason: e.to_string(),
+                    }
+                } else {
+                    let inner = e.downcast::<FetchError>().unwrap();
```
Comment (bnjjj) [inline apollo-router/src/query_planner/fetch.rs:277]: You should not call `.unwrap()` but just match over the `Option` instead.

File: apollo-router/src/query_planner/fetch.rs [hunk 2]
```diff
@@ -266,10 +266,30 @@ impl FetchNode {
...
+                                reason: inner.to_string(),
+                            }
+                        }
+                        _ => FetchError::SubrequestHttpError {
+                            status_code: None,
+                            service: service_name.to_string(),
+                            reason: inner.to_string(),
+                        },
+                    }
+                }
```
Comment (bnjjj) [inline apollo-router/src/query_planner/fetch.rs:292]: I think it's the same than just doing (but maybe I'm missing something):
```rust
FetchError::SubrequestHttpError {
  status_code,
  service: service_name.to_string(),
  reason: inner.to_string(),
}
```

File: apollo-router/src/query_planner/fetch.rs (hunk 1 above)
Comment (scottdouglas1989) [reply apollo-router/src/query_planner/fetch.rs]: Good catch! That is correct.

File: apollo-router/src/query_planner/fetch.rs (hunk 2 above)
Comment (scottdouglas1989) [reply apollo-router/src/query_planner/fetch.rs]: With the match case on the downcast from your previous comment, I was able to simplify even further, removing the `is` check and simply pushing forward the existing `FetchError::SubrequestHttpError` (WITH `status_code` already included) if we find a match. Thanks for the feedback! This removes a copy as well from the previous implementation 💯

**Review by abernix — DISMISSED** [review]
Parameter shape and naming-wise, this doesn't quite seem aligned with [the behavior of Apollo Server](https://www.apollographql.com/docs/apollo-server/data/errors/#setting-http-status-code-and-headers).  Should it be?  My hunch is to believe the answer should be yes to avoid additional surprise for clients.

**Comment by bnjjj** [comment]
I pushed a small change to follow the apollo server conventions. we will now have

```
"errors": [
		{
			"message": "HTTP fetch failed from 'players': HTTP fetch failed from 'players': 429: Too Many Requests",
			"path": [],
			"extensions": {
				"http": {
                                     "status": 429,
                                },
				"service": "players",
				"reason": "HTTP fetch failed from 'players': 429: Too Many Requests",
				"code": "SUBREQUEST_HTTP_ERROR"
			}
		}
	]
```

does it look good to you @scottdouglas1989 ?

**Comment by scottdouglas1989** [comment]
Looks great guys! Thanks for tackling that, and thanks @abernix for the additional feedback 💯

**Pull request merged**